### DISCUSSION TITLE
ARM64 version

### DISCUSSION THREAD

**Discussion by saikunjp** [pybamm-team/PyBaMM#2434] [General]
Hello,

Is there a version of the program compiled for Apple's M1/M2 processor?

Thanks

**Comment by Saransh-cpp** [comment]
Hi, @saikunjp!

There are some confusions related to PyBaMM's M1/M2 support -
- `PyBaMM` supports both `M1` and `M2` if both `casadi` and `PyBaMM` are installed from the source (confirmed locally)
- `PyBaMM` should support `M1` and `M2` chips with `rosetta` installed on your system (cc: @tinosulzer)

The confusion originates from the fact that `casadi` [does not support](https://github.com/casadi/casadi/issues/2829) M1/M2 chips.

Writing down the "build from source" steps for `casadi` and `PyBaMM` -
>**Note**
> You don't have to follow these if PyBaMM works with rosetta installed on your machine. Go through these steps only if `pip3 install pybamm` does not work after installing rosetta -

**Install casadi**
```
brew install cmake swig
git clone https://github.com/casadi/casadi.git casadi
cd casadi
git checkout 3.5.5
mkdir build
cd build
cmake -DWITH_PYTHON=ON \
-DWITH_PYTHON3=ON \
-DPYTHON_INCLUDE_DIR=$(python3 -c "from distutils.sysconfig import get_python_inc; print(get_python_inc())") \
-DPYTHON_LIBRARY=$(python3 -c "import distutils.sysconfig as sysconfig; print(sysconfig.get_config_var('LIBDIR'))") \
..
make
sudo make install
```

**Install PyBaMM**
Make sure you are outside of the `casadi` folder.
```
git clone https://github.com/pybamm-team/PyBaMM PyBaMM
cd PyBaMM
git checkout 22.10  # replace with the latest version
```
Comment out the following line from the `setup.py` file of the locally cloned `PyBaMM` -
https://github.com/pybamm-team/PyBaMM/blob/bc83ce7cc838dac36946574950e807d788cf7af7/setup.py#L198
then follow the instructions available here - https://pybamm.readthedocs.io/en/latest/install/install-from-source.html!

**Reply by saikunjp** [reply]
Thanks so much for your reply!  
Compiling from source is not easy :).  I got errors when installing scikits/odes (following the instructions on the webpage of PyBamm to use tor), the 13 errors are all of the type:
       scikits/odes/sundials/cvode.c:24010:91: error: too few arguments to function call, expected 3, have 2
              __pyx_v_atol = N_VMake_Serial(__pyx_v_self->N, ((realtype *)__pyx_v_np_atol->data));
I think it has to do with the version of Sundials but could not find a solution yet.

      error: Command "clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/TOM/opt/anaconda3/include -arch arm64 -I/Users/TOM/opt/anaconda3/include -fPIC -O2 -isystem /Users/TOM/opt/anaconda3/include -arch arm64 -I/opt/homebrew/opt/openblas/include -ftrapping-math -I/opt/homebrew/Cellar/sundials/6.3.0/include -I/private/var/folders/bp/d_k8v4zs1k9d6z6phjwkf6n00000gn/T/pip-build-env-hedi9h2y/overlay/lib/python3.9/site-packages/numpy/core/include -Ibuild/src.macosx-11.1-arm64-3.9/numpy/distutils/include -I/Users/TOM/workspace/PyBaMM/.tox/dev/include -I/Users/TOM/opt/anaconda3/include/python3.9 -c scikits/odes/sundials/cvode.c -o build/temp.macosx-11.1-arm64-cpython-39/scikits/odes/sundials/cvode.o -MMD -MF build/temp.macosx-11.1-arm64-cpython-39/scikits/odes/sundials/cvode.o.d" failed with exit status 1

**Reply by Saransh-cpp** [reply]
Yes, installation from the source can be frustrating at times 😅

Could you try not to install `scikits.odes` as a start? I think `tox` will automatically trigger this installation. You could do this to avoid using `tox` -

```
cd PyBaMM
python3 -m venv .env
. .env/bin/activate
python3 -m pip install .
```

(You can always add these optional solvers after the initial installation.)
